opencxl/apps/cxl_image_classification_host.py

```diff
@@ -103,30 +112,100 @@ async def write_mmio(self, address: int, size: int, value: int):
     async def read_mmio(self, address: int, size: int) -> int:
         return await self._root_complex.read_mmio(address, size)
 
-    async def write_cxl_mem(self, address: int, size: int, value: int):
-        await self._root_complex.write_cxl_mem(address, size, value)
+    def to_device_mmio_addr(self, device: int, addr: int) -> int:
+        assert addr < self._dev_mmio_ranges[device][1]
+        return self._dev_mmio_ranges[device][0] + addr
 
-    async def read_cxl_mem(self, address: int, size: int) -> int:
-        return await self._root_complex.read_cxl_mem(address, size)
+    def to_device_mem_addr(self, device: int, addr: int) -> int:
+        assert addr < self._dev_mem_ranges[device][1]
+        return self._dev_mem_ranges[device][0] + addr
 
-    def to_device_addr(self, device: int, addr: int) -> int:
-        unaligned_offset = addr % self._interleave_gran
-        dev_offset = (addr - unaligned_offset) * self._device_count + (
-            device * self._interleave_gran
+    async def _host_process_validation_type1(self):
+        categories = glob.glob(self._train_data_path + "/val/*")
+        self._total_samples = len(categories) * self._sample_from_each_category
+        self._validation_results: List[List[Dict[str, float]]] = [[] for _ in self._total_samples]
+        pic_id = 0
+        pic_data_mem_loc = 0x00008000
+        for c in categories:
+            category_pics = glob.glob(f"{c}/*.JPEG")
+            sample_pics = sample(category_pics, self._sample_from_each_category)
+            category_name = c.split(os.path.sep)[-1]
+            self._sampled_file_categories += [category_name] * self._sample_from_each_category
+            for s in sample_pics:
+                with open(s, "rb") as f:
+                    pic_data = f.read()
+                    pic_data_int = int.from_bytes(pic_data, "little")
+                    pic_data_len = len(pic_data)
+                    await self.store(pic_data_mem_loc, pic_data_len, pic_data_int)
+                    for dev in range(self._device_count):
```

Comment:
`dev_idx`? `idx`? `dev_num`?

`dev` sounds like it should be an object. Please change this `_save_validation_result_type1()` (and any other place) as well.